lib/boyer_moore.rb

```diff
@@ -1,15 +1,34 @@
 require_relative "./boyer_moore/version"
 
 module BoyerMoore
-  def self.search(haystack, needle_string)
-    needle = Needle.new(needle_string)
 
-    haystack_index = 0
-    while haystack_index <= haystack.size - needle.size
-      if skip_by = needle.match_or_skip_by(haystack, haystack_index)
-        haystack_index += skip_by
-      else
-        break haystack_index # Found a match at haystack_index!
+  # Yields the index of each match of needle_string in haystack.
+  # If no block given, returns an Enumerator that does the same.
+  # Starts at starting_haystack_index (default = 0).
+  def self.search(haystack, needle_string, starting_haystack_index = 0, &block)
+    enumerator = Enumerator.new(haystack, needle_string, starting_haystack_index)
+    if block
+      enumerator.each(&block)
+    else
+      enumerator.to_enum(:each)
+    end
+  end
+
+  class Enumerator
```

Comment:
Perhaps YAGNI, but you could `include Enumerable` to gain other methods like `map` or `any?`.